Google Maps

Allcharles opened the pull request:
Added google maps to project and site pages. Zoom level on the map may need some tweaking later 

New dependencies:
- [@agm/core](https://github.com/SebastianM/angular-google-maps)
- [snazzy-info-window](https://github.com/SebastianM/angular-google-maps)

Closes #10 

Edit # 1: Updated snazzy-info-window link
Edit # 2: Example to show off update
![image](https://user-images.githubusercontent.com/3955116/70485398-9203d800-1b3a-11ea-93d3-7f22a6a30d1a.png)

Allcharles committed:
Added map to project details page

Allcharles committed:
Working project details map

Allcharles committed:
Added Snazzy Info Window

Allcharles committed:
Added maps to site details

atruskie commented on src/app/component/projects/pages/details/details.component.html:
```diff
@@ -56,18 +57,46 @@ <h5>
 <ng-container *ngIf="state === 'notFound'">
   <app-item-not-found></app-item-not-found>
 </ng-container>
 
```
It seems strange these are part of the details view. Do they have to be specified for every view? Is there a more general way of doing this?

Allcharles replied:
I would almost prefer a more general way of performing this. Perhaps a component which takes a state variable as an input?

atruskie replied:
So when errors like this happen, the routing system should render an error component in place of the component that was meant to be rendered.

E.g. 404 on projects/123 should render <not-found /> in place of <project-details />

This should be a separate issue, let's not fix it here. Create an issue, please.

Allcharles replied:
Issue created here #30.

atruskie commented on src/app/component/projects/pages/details/details.component.ts:
```diff
@@ -54,6 +54,8 @@ export class DetailsComponent extends PageComponent implements OnInit {
   sites: Site[];
   errorCodes = this.sitesApi.apiReturnCodes;
   state = "loading";
+  avgLat = 0;
```
don't abbreviate names

Allcharles replied:
Resolved: Removed variables

atruskie commented on src/app/component/projects/pages/details/details.component.ts:
```diff
@@ -82,6 +84,13 @@ export class DetailsComponent extends PageComponent implements OnInit {
+            });
+            this.avgLat /= sites.length;
+            this.avgLong /= sites.length;
+
```
Averaging of lat/long doesn't work like this.

But more so, why are we averaging?

Allcharles replied:
We need the average latitude and longitude for the `<agm-map>` component as it does not have an inbuilt averaging feature (as far as I tested, will look over again). I will look into a better implementation of this calculation.

atruskie replied:
No, please do no implement such an algorithm.

But what is being averaged here? Are we averaging the site locations? Why?

Previous maps just showed all sites in the project details page - why not again?

Allcharles replied:
This is to average the site locations, its needed because you need to supply the `<agm-map>` component with a longitude and latitude to focus the map on. Adding pins has no affect on the zoom level or position of the map. Below is an image of the map that is generated. The pins are added using a for loop in the html which iterates over the sites and adds pins, snazzy-info-window creates the white background for the pin label, and the average latitude and longitude focus the map to where the pins are located.
![image](https://user-images.githubusercontent.com/3955116/67116872-a0d8a900-f1e1-11e9-8008-f557691f4e13.png)

Allcharles replied:
Resolved: Found how to use the agm/core library to auto focus on the pins

atruskie replied:
Fantastic 👍

atruskie reviewed (commented):
I'd like to see if what the snazzy dep offers, and understand what the averaging is about.

But otherwise, nice, and clean ✔

Allcharles committed:
Removed averaging of longitude and latitude

Fixed minor aethetic bug where map min-height of 400px was not functional.

Allcharles committed:
Fixed map component unit test imports

Allcharles committed:
Merge branch 'master' into maps

Allcharles committed:
Added project details unit tests

- Renamed cards.module file
- Updated typescript definition files
- Updated mock service extensions
- Fixed fontawesome error during unit tests
- Created mock google maps component

Allcharles committed:
Cleanup site details component code

Allcharles committed:
Created site details unit tests

Allcharles committed:
Fixed broken tests

Allcharles committed:
Added site-card component unit tests

Allcharles committed:
Fixed bug in createMap function
